airbyte-cdk/bulk/toolkits/load-object-storage/src/main/kotlin/io/airbyte/cdk/load/file/object_storage/ObjectStorageFormattingWriter.kt

```diff
@@ -155,11 +169,73 @@ class ParquetFormattingWriter(
     }
 
     override fun accept(record: DestinationRecord) {
-        val dataMapped =
-            pipeline
-                .map(record.data, record.meta?.changes)
-                .withAirbyteMeta(stream, record.emittedAtMs, rootLevelFlattening)
-        writer.write(dataMapped.toAvroRecord(mappedSchema, avroSchema))
+        val dataMapped = pipeline.map(record.data, record.meta?.changes)
+        numCapturedChanges.addAndGet(dataMapped.second.size.toLong())
+        val withMeta = dataMapped.withAirbyteMeta(stream, record.emittedAtMs, rootLevelFlattening)
+        writer.write(withMeta.toAvroRecord(mappedSchema, avroSchema))
+    }
+
+    override fun close() {
+        writer.close()
+    }
+}
+
+@Singleton
+@Secondary
+class BufferedFormattingWriterFactory<T : OutputStream>(
+    private val writerFactory: ObjectStorageFormattingWriterFactory,
+    private val formatConfigProvider: ObjectStorageFormatConfigurationProvider,
+    private val compressionConfigurationProvider: ObjectStorageCompressionConfigurationProvider<T>,
+    private val uploadConfigProvider: ObjectStorageUploadConfigurationProvider
+) {
+    fun create(stream: DestinationStream): BufferedFormattingWriter<T> {
+        val outputStream = ByteArrayOutputStream()
+        val processor =
+            compressionConfigurationProvider.objectStorageCompressionConfiguration.compressor
+        val wrappingBuffer = processor.wrapper.invoke(outputStream)
+        val writer = writerFactory.create(stream, wrappingBuffer)
+        return BufferedFormattingWriter(
+            writer,
+            uploadConfigProvider.objectStorageUploadConfiguration.streamingUploadPartSize,
+            outputStream,
+            processor,
+            wrappingBuffer
+        )
+    }
+}
+
+class BufferedFormattingWriter<T : OutputStream>(
+    private val writer: ObjectStorageFormattingWriter,
+    private val partSize: Long,
+    private val buffer: ByteArrayOutputStream,
+    private val streamProcessor: StreamProcessor<T>,
+    private val wrappingBuffer: T
+) : ObjectStorageFormattingWriter {
+    override val numCapturedChanges: AtomicLong = writer.numCapturedChanges
+
+    override fun accept(record: DestinationRecord) {
+        writer.accept(record)
+    }
+
+    fun isDataSufficient(): Boolean {
+        return buffer.size() >= partSize
+    }
```

Comment:
Nit: this seems orthogonal to the buffered writer. Since it's called strictly externally, we can easily pull this out.

Reply:
Yeah, it should just be a size check.